src/main/java/com/google/cloud/spanner/pgadapter/ConnectionHandler.java

```diff
@@ -713,23 +724,52 @@ public void setWellKnownClient(WellKnownClient wellKnownClient) {
    * executed.
    */
   public void maybeDetermineWellKnownClient(Statement statement) {
-    if (!this.hasDeterminedClientUsingQuery
-        && this.wellKnownClient == WellKnownClient.UNSPECIFIED
-        && getServer().getOptions().shouldAutoDetectClient()) {
-      this.wellKnownClient = ClientAutoDetector.detectClient(ImmutableList.of(statement));
-      if (this.wellKnownClient != WellKnownClient.UNSPECIFIED) {
-        logger.log(
-            Level.INFO,
-            () ->
-                String.format(
-                    "Well-known client %s detected for connection %d.",
-                    this.wellKnownClient, getConnectionId()));
+    if (!this.hasDeterminedClientUsingQuery) {
+      if (this.wellKnownClient == WellKnownClient.UNSPECIFIED
+          && getServer().getOptions().shouldAutoDetectClient()) {
+        setWellKnownClient(ClientAutoDetector.detectClient(ImmutableList.of(statement)));
+      }
+      maybeSetApplicationName();
+    }
+    // Make sure that we only try to detect the client once.
+    this.hasDeterminedClientUsingQuery = true;
+  }
+
+  /**
+   * This is called by the extended query protocol {@link
+   * com.google.cloud.spanner.pgadapter.wireprotocol.ParseMessage} to give the connection the
+   * opportunity to determine the client that is connected based on the data in the (first) parse
+   * messages.
+   */
+  public void maybeDetermineWellKnownClient(ParseMessage parseMessage) {
+    if (!this.hasDeterminedClientUsingQuery) {
+      if (this.wellKnownClient == WellKnownClient.UNSPECIFIED
+          && getServer().getOptions().shouldAutoDetectClient()) {
+        setWellKnownClient(ClientAutoDetector.detectClient(parseMessage));
       }
+      maybeSetApplicationName();
     }
     // Make sure that we only try to detect the client once.
     this.hasDeterminedClientUsingQuery = true;
   }
 
+  private void maybeSetApplicationName() {
+    if (this.wellKnownClient != WellKnownClient.UNSPECIFIED
+        && getExtendedQueryProtocolHandler() != null
+        && Strings.isNullOrEmpty(
+            getExtendedQueryProtocolHandler()
+                .getBackendConnection()
```

Comment:
Wondering if anything can return null in this chain of calls?

Reply:
Good point. Just verified that:
- ExtendedQueryProtocolHandler.backendConnection is final and initialised to a non-null in all constructors.
- BackendConnection.sessionState is final and initialised to a non-null value in all constructors.
- SessionState.get(String, String) is guaranteed to return a non-null value or throw an exception in case of an undefined parameter name (`application_name` is a fixed parameter name that should always be known).

I've added an extra safeguard in the form of a try-catch-ignore to account for the theoretical possibility that we allow users to supply their own pg_settings.txt file in the future, and that anyone removes the `application_name` setting from that file.